SourceX/qol.cpp

```diff
@@ -0,0 +1,460 @@
+		}
+	}
+
+	for (int j = 0; j < borderWidth; j++) {
+		for (int i = -xOffset - (fillCorners ? borderWidth : 0); i < width + xOffset + (fillCorners ? borderWidth : 0); i++) {
+			ENG_set_pixel(xPos + i, yPos + j - yOffset - borderWidth, borderColor);
+		}
+		for (int i = -xOffset; i < width + xOffset + (fillCorners ? borderWidth : 0); i++) {
+			ENG_set_pixel(xPos + i, yPos + j + yOffset + height, borderColor);
+		}
+	}
+
+	for (int j = -yOffset; j < height + yOffset + (fillCorners ? borderWidth : 0); j++) {
+		for (int i = 0; i < borderWidth; i++) {
+			ENG_set_pixel(xPos + i - xOffset - borderWidth, yPos + j, borderColor);
+			ENG_set_pixel(xPos + i + xOffset + width, yPos + j, borderColor);
+		}
+	}
+
+	bool drawImmu = false;
+	int immuOffset = 0 + GetTextWidth(immuText) - 5;
+	for (int k = 0; k < resSize; ++k) {
+		if (mres & immunes[k]) {
+			drawImmu = true;
+			for (int j = 0; j < square; j++) {
+				for (int i = 0; i < square; i++) {
+					ENG_set_pixel(xPos + i + immuOffset, yPos + height + j + yOffset + borderWidth + 2 - 15, resistColors[k]);
+				}
+			}
+			immuOffset += square + 2;
+		}
+	}
+
+	int newX = xPos + BORDER_LEFT;
+	//int newY = yPos + height - 3;
+
+	char text[166];
+	strcpy(text, mon->mName);
+	if (mon->leader > 0)
+		strcat(text, " (minion)");
+
+	int namecolor = COL_WHITE;
+	if (specialMonster)
+		namecolor = COL_GOLD;
+	PrintGameStr(newX - GetTextWidth(text) / 2, 30, text, namecolor);
+	PrintGameStr(newX - GetTextWidth("/") / 2, 43, "/", COL_WHITE);
+
+	sprintf(text, "%d", (maxLife >> 6));
+	PrintGameStr(newX + GetTextWidth("/"), 43, text, COL_WHITE);
+
+	sprintf(text, "%d", (currentLife >> 6));
+	PrintGameStr(newX - GetTextWidth(text) - GetTextWidth("/"), 43, text, COL_WHITE);
+
+	PrintGameStr(newX - width / 2, 59, resText, COL_GOLD);
+
+	sprintf(text, "kills: %d", monstkills[mon->MType->mtype]);
+	PrintGameStr(newX - GetTextWidth("kills:") / 2 - 30, 59, text, COL_WHITE);
+
+	if (drawImmu)
+		PrintGameStr(newX - width / 2, 46, immuText, COL_GOLD);
+
+	PrintGameStr(newX + width / 2 - GetTextWidth(vulnText), 59, vulnText, COL_RED);
+}
+
+void DrawXPBar()
+{
+	if (!drawXPBar)
+		return;
+	int barSize = 306; // *ScreenWidth / 640;
+	int offset = 3;
+	int barRows = 3; // *ScreenHeight / 480;
+	int dividerHeight = 3;
+	int numDividers = 10;
+	int barColor = 242; /*242white, 142red, 200yellow, 182blue*/
+	int emptyBarColor = 0;
+	int frameColor = 242;
+	int yPos = BORDER_TOP + SCREEN_HEIGHT - 8;
+
+	PrintGameStr(145 + (SCREEN_WIDTH - 640) / 2, SCREEN_HEIGHT - 4, "XP", COL_WHITE);
+	int charLevel = plr[myplr]._pLevel;
+	if (charLevel != MAXCHARLEVEL - 1) {
+		int curXp = ExpLvlsTbl[charLevel];
+		int prevXp = ExpLvlsTbl[charLevel - 1];
+		int prevXpDelta = curXp - prevXp;
+		int prevXpDelta_1 = plr[myplr]._pExperience - prevXp;
+		if (plr[myplr]._pExperience >= prevXp) {
+			int visibleBar = barSize * (unsigned __int64)prevXpDelta_1 / prevXpDelta;
+
+			for (int i = 0; i < visibleBar; ++i) {

```

Comment:
You may invert the loop for perf

Reply:
code readability suffers from that and I gained 0 perf from doing that - modern compilers probably apply that by default or it's simply not worth it